File: week-08/rimrim990/보석 쇼핑.py
def solution(gems):
  cnt = {}
  for gem in gems:
    cnt[gem] = 0

  bag = {gems[0]}
  cnt[gems[0]] += 1
  answer = [1, len(gems)]

  end = 1
  for start in range(len(gems)):
    while len(bag) < len(cnt) and end < len(gems):
      bag.add(gems[end])
      cnt[gems[end]] += 1
      end += 1

    if len(bag) == len(cnt) and end-1-start < answer[1]-answer[0]:
      answer = [start+1, end]

    cnt[gems[start]] -= 1
    if cnt[gems[start]] == 0:
      bag.remove(gems[start])

  return answer

File: week-08/dahyen0o/보석-쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(final String[] gems) {
        final Map<String, Integer> counts = new HashMap<>();
        for(String gem : gems) {
            counts.putIfAbsent(gem, 0);
        }

        if(counts.size() == 1) {
            return new int[] {1, 1};
        }

        final Set<String> contains = new HashSet<>();

        final int[] answer = {0, Integer.MAX_VALUE};
        int left = 0, right = 0;
        counts.compute(gems[0], (key, value) -> value + 1);
        contains.add(gems[0]);

        while(right < gems.length - 1) {
            if(contains.size() == counts.size() && answer[1] - answer[0] > right - left) {
                answer[0] = left;
                answer[1] = right;
            }

            if(contains.size() == counts.size()) {
                counts.compute(gems[left], (key, value) -> value - 1);
                if(counts.get(gems[left]) == 0) {
                    contains.remove(gems[left]);
                }
                left++;
            } else {
                right++;
                counts.compute(gems[right], (key, value) -> value + 1);
                contains.add(gems[right]);
            }
        }

        while(left < right) {
            if(contains.size() == counts.size() && answer[1] - answer[0] > right - left) {
                answer[0] = left;
                answer[1] = right;
            }

            counts.compute(gems[left], (key, value) -> value - 1);
            if(counts.get(gems[left]) == 0) {
                contains.remove(gems[left]);
            }
            left++;
        }

        answer[0]++;
        answer[1]++;
        return answer;
    }
}
